src/natcap/invest/forest_carbon_edge_effect.py

```diff
@@ -612,20 +612,48 @@ def _map_distance_from_tropical_forest_edge(
         base_lulc_raster_path)['nodata']
 
     def mask_non_forest_op(lulc_array):
-        """Converts forest lulc codes to 1."""
-        non_forest_mask = ~numpy.in1d(
-            lulc_array.flatten(), forest_codes).reshape(lulc_array.shape)
+        """Convert forest lulc codes to 0.
+        Args:
+            lulc_array (numpy.ndarray): array representing a LULC raster where
+                each forest LULC code is in `forest_codes`.
+        Returns:
+            numpy.ndarray with the same shape as lulc_array. All pixels are
+                0 (forest), 1 (non-forest), or 255 (nodata).
+        """
+        non_forest_mask = ~numpy.isin(lulc_array, forest_codes)
         nodata_mask = lulc_array == lulc_nodata
+        # where LULC has nodata, set value to forest_mask_nodata value (255)
+        # where LULC has data, set to 0 if LULC is a forest type, 1 if it's not
         return numpy.where(nodata_mask, forest_mask_nodata, non_forest_mask)
 
     pygeoprocessing.raster_calculator(
         [(base_lulc_raster_path, 1)], mask_non_forest_op,
         target_non_forest_mask_path, gdal.GDT_Byte, forest_mask_nodata)
 
     # Do the distance transform on non-forest pixels
+    # This is the distance from each pixel to the nearest pixel with value 1.
+    #   - for forest pixels, this is the distance to the forest edge
+    #   - for non-forest pixels, this is 0
+    #   - for nodata pixels, distance is calculated but is meaningless
     pygeoprocessing.distance_transform_edt(
         (target_non_forest_mask_path, 1), edge_distance_path)
 
+    # mask out the meaningless distance pixels so they don't affect the output
+    edge_distance_raster = gdal.OpenEx(edge_distance_path, gdal.GA_Update)
+    edge_distance_band = edge_distance_raster.GetRasterBand(1)
+    for (lulc_offset, lulc_block), (_, distance_block) in zip(
+            pygeoprocessing.iterblocks(
+                (base_lulc_raster_path, 1), largest_block=2**12),
+            pygeoprocessing.iterblocks(
+                (edge_distance_path, 1), largest_block=2**12)):
+        # where LULC has nodata, overwrite edge distance with nodata value
+        masked_distance_block = numpy.where(
+            lulc_block == lulc_nodata, -1, distance_block)
+        edge_distance_band.WriteArray(
```

Comment:
Is there anything wrong with alternately reading and writing to a raster like this? Because it's reading one block, then overwriting that block, then reading the next block, I think it shouldn't conflict.

Reply:
That's a good question. I think it'd be fine. BUT, if you know the rasters are aligned you could do:
```python

base_lulc_raster = gdal.OpenEx(base_lulc_raster_path)      
base_lulc_band = base_lulc_raster.GetRasterBand(1)         
                                                  
edge_distance_raster = gdal.OpenEx(edge_distance_raster_path)
edge_distance_band = edge_distance_raster.GetRasterBand(1)   

for offset_dict in pygeoprocessing.iterblocks(
    (base_lulc_raster_path, 1), offset_only=True):

     lulc_block = lulc_band.ReadAsArray(**offset_dict)
     edge_distance_block = edge_distance_band.ReadAsArray(**offset_dict)
```

BUT, it might also just be better to call another `raster_calculator` call too, with a simple op that masks the nodata. This would require a temporary file of sorts for that intermediate step.

Reply:
I think I like your code snippet the best, it seems the most simple.